src/main/java/com/google/api/generator/gapic/protowriter/Writer.java

```diff
@@ -43,52 +45,104 @@ public static CodeGeneratorResponse write(
       String outputFilePath) {
     ByteString.Output output = ByteString.newOutput();
     JavaWriterVisitor codeWriter = new JavaWriterVisitor();
-    JarOutputStream jos = null;
+    JarOutputStream jos;
     try {
       jos = new JarOutputStream(output);
     } catch (IOException e) {
       throw new GapicWriterException(e.getMessage());
     }
 
     for (GapicClass gapicClazz : clazzes) {
-      ClassDefinition clazz = gapicClazz.classDefinition();
+      String classPath = writeClazz(gapicClazz, codeWriter, jos);
+      writeSamples(gapicClazz, getSamplePackage(gapicClazz), classPath, jos);
+    }
+
+    writeMetadataFile(context, writePackageInfo(gapicPackageInfo, codeWriter, jos), jos);
+
+    try {
+      jos.finish();
+      jos.flush();
+    } catch (IOException e) {
+      throw new GapicWriterException(e.getMessage());
+    }
+
+    CodeGeneratorResponse.Builder response = CodeGeneratorResponse.newBuilder();
+    response
+        .setSupportedFeatures(CodeGeneratorResponse.Feature.FEATURE_PROTO3_OPTIONAL_VALUE)
+        .addFileBuilder()
+        .setName(outputFilePath)
+        .setContentBytes(output.toByteString());
+    return response.build();
+  }
 
-      clazz.accept(codeWriter);
-      String code = codeWriter.write();
-      codeWriter.clear();
+  private static String writeClazz(
+      GapicClass gapicClazz, JavaWriterVisitor codeWriter, JarOutputStream jos) {
+    ClassDefinition clazz = gapicClazz.classDefinition();
+
+    clazz.accept(codeWriter);
+    String code = codeWriter.write();
+    codeWriter.clear();
+
+    String path = getPath(clazz.packageString(), clazz.classIdentifier().name());
+    String className = clazz.classIdentifier().name();
+    JarEntry jarEntry = new JarEntry(String.format("%s/%s.java", path, className));
+    try {
+      jos.putNextEntry(jarEntry);
+      jos.write(code.getBytes(StandardCharsets.UTF_8));
+    } catch (IOException e) {
+      throw new GapicWriterException(
+          String.format(
+              "Could not write code for class %s.%s: %s",
+              clazz.packageString(), clazz.classIdentifier().name(), e.getMessage()));
+    }
+    return path;
+  }
 
-      String path = getPath(clazz.packageString(), clazz.classIdentifier().name());
-      String className = clazz.classIdentifier().name();
-      JarEntry jarEntry = new JarEntry(String.format("%s/%s.java", path, className));
+  private static void writeSamples(
+      GapicClass gapicClazz, String pakkage, String clazzPath, JarOutputStream jos) {
+    for (Sample sample : gapicClazz.samples()) {
+      JarEntry jarEntry =
+          new JarEntry(
+              String.format(
+                  "samples/snippets/generated/%s/%s/%s/%s.java",
+                  clazzPath,
+                  sample.regionTag().serviceName().toLowerCase(),
+                  sample.regionTag().rpcName().toLowerCase(),
```

Comment:
Can you please check Vadym's recent PR regarding empty services? With his code change and your code change here, we are going to create an empty Sample file under the path /empty, which I think may not be a good idea going forward. I think we need to either filter out empty services here or when creating Samples, unless you think empty sample is a valid case.

Reply:
I replied to the PR. I don't think empty should be passed in. There's a reason a region tag can't be built without rpcName.